Implement `std::fmt::Display` for `Query`

It might be helpful to log the queries. 
Currently, it is not possible:
```
error[E0277]: `Query` doesn't implement `std::fmt::Display`
  |
8 |     println!("{query}");
  |               ^^^^^^^ `Query` cannot be formatted with the default formatter
  |
  = help: the trait `std::fmt::Display` is not implemented for `Query`
```

The current implementation of SQLBuilder moves `mut self,` so it looks like some refactoring will be required there.

--- a/src/query.rs
+++ b/src/query.rs
@@ -1,6 +1,6 @@
 use hyper::{header::CONTENT_LENGTH, Method, Request};
 use serde::Deserialize;
-use std::fmt;
+use std::fmt::Display;
 use url::Url;
 
 use crate::headers::with_request_headers;
@@ -23,18 +23,17 @@ pub struct Query {
     sql: SqlBuilder,
 }
 
-impl fmt::Debug for Query {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{sql}", sql = self.sql)
-    }
-}
-
 impl Query {
     pub(crate) fn new(client: &Client, template: &str) -> Self {
         Self {
             client: client.clone(),
             sql: SqlBuilder::new(template),
         }
+    }
+
+    /// Display SQL query as string.
+    pub fn sql_display(&self) -> &impl Display {
+        &self.sql
     }
 
     /// Binds `value` to the next `?` in the query.

--- a/src/query.rs
+++ b/src/query.rs
@@ -1,5 +1,6 @@
 use hyper::{header::CONTENT_LENGTH, Method, Request};
 use serde::Deserialize;
+use std::fmt;
 use url::Url;
 
 use crate::headers::with_request_headers;
@@ -20,6 +21,12 @@ const MAX_QUERY_LEN_TO_USE_GET: usize = 8192;
 pub struct Query {
     client: Client,
     sql: SqlBuilder,
+}
+
+impl fmt::Debug for Query {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{sql}", sql = self.sql)
+    }
 }
 
 impl Query {

--- a/src/sql/mod.rs
+++ b/src/sql/mod.rs
@@ -1,4 +1,4 @@
-use std::fmt::Display;
+use std::fmt::{self, Display};
 
 use crate::{
     error::{Error, Result},
@@ -11,17 +11,36 @@ mod bind;
 pub(crate) mod escape;
 mod ser;
 
-#[derive(Clone)]
+#[derive(Debug, Clone)]
 pub(crate) enum SqlBuilder {
     InProgress(Vec<Part>),
     Failed(String),
 }
 
-#[derive(Clone)]
+#[derive(Debug, Clone)]
 pub(crate) enum Part {
     Arg,
     Fields,
     Text(String),
+}
+
+/// Display SQL query as string.
+impl fmt::Display for SqlBuilder {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            SqlBuilder::InProgress(parts) => {
+                for part in parts {
+                    match part {
+                        Part::Arg => write!(f, "?")?,
+                        Part::Fields => write!(f, "?fields")?,
+                        Part::Text(text) => write!(f, "{text}")?,
+                    }
+                }
+            }
+            SqlBuilder::Failed(err) => write!(f, "{err}")?,
+        }
+        Ok(())
+    }
 }
 
 impl SqlBuilder {
@@ -141,9 +160,29 @@ mod tests {
     #[test]
     fn bound_args() {
         let mut sql = SqlBuilder::new("SELECT ?fields FROM test WHERE a = ? AND b < ?");
+        assert_eq!(
+            sql.to_string(),
+            "SELECT ?fields FROM test WHERE a = ? AND b < ?"
+        );
+
         sql.bind_arg("foo");
+        assert_eq!(
+            sql.to_string(),
+            "SELECT ?fields FROM test WHERE a = 'foo' AND b < ?"
+        );
+
         sql.bind_arg(42);
+        assert_eq!(
+            sql.to_string(),
+            "SELECT ?fields FROM test WHERE a = 'foo' AND b < 42"
+        );
+
         sql.bind_fields::<Row>();
+        assert_eq!(
+            sql.to_string(),
+            "SELECT `a`,`b` FROM test WHERE a = 'foo' AND b < 42"
+        );
+
         assert_eq!(
             sql.finish().unwrap(),
             r"SELECT `a`,`b` FROM test WHERE a = 'foo' AND b < 42"

--- a/tests/it/query.rs
+++ b/tests/it/query.rs
@@ -221,7 +221,7 @@ async fn prints_query() {
 
     let q = client.query("SELECT ?fields FROM test WHERE a = ? AND b < ?");
     assert_eq!(
-        format!("{q:?}"),
+        format!("{}", q.sql_display()),
         "SELECT ?fields FROM test WHERE a = ? AND b < ?"
     );
 }

--- a/tests/it/query.rs
+++ b/tests/it/query.rs
@@ -214,3 +214,14 @@ async fn overrides_client_options() {
     // should override the client options
     assert_eq!(value, override_value);
 }
+
+#[tokio::test]
+async fn prints_query() {
+    let client = prepare_database!();
+
+    let q = client.query("SELECT ?fields FROM test WHERE a = ? AND b < ?");
+    assert_eq!(
+        format!("{q:?}"),
+        "SELECT ?fields FROM test WHERE a = ? AND b < ?"
+    );
+}

